In [ ]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

/kaggle/input/ultimate-spotify-tracks-db/SpotifyFeatures.csv


# SPOTIFY DATASET BASED MUSIC RECOMMENDATION SYSTEM

 Built a music recommendation engine on 230,000+ Spotify tracks using cosine similarity across audio features like danceability, energy, and valence. Implemented dual recommendation modes (content-based and artist-based) 

 **Author :** Sai Chaithanyia P

### Loading the Dataset

The Spotify dataset is loaded into a Pandas DataFrame.  
This dataset contains track-level metadata and audio features required for building a content-based recommendation system.  
Initial inspection ensures the dataset is correctly loaded.


In [ ]:
df=pd.read_csv("data/SpotifyFeatures.csv")

### Dataset Inspection

Displays the first few rows of the dataset to understand feature structure, column names, and data types.  
This step helps verify that all relevant attributes are present before preprocessing.


In [3]:
df.head()

,genre,artist_name,track_name,track_id,popularity,acousticness,danceability,duration_ms,energy,instrumentalness,key,liveness,loudness,mode,speechiness,tempo,time_signature,valence
0,Movie,Henri Salvador,C'est beau de faire un Show,0BRjO6ga9RKCKjfDqeFgWV,0,0.611,0.389,99373,0.910,0.000,C#,0.3460,-1.828,Major,0.0525,166.969,4/4,0.814
1,Movie,Martin & les fées,Perdu d'avance (par Gad Elmaleh),0BjC1NfoEOOusryehmNudP,1,0.246,0.590,137373,0.737,0.000,F#,0.1510,-5.559,Minor,0.0868,174.003,4/4,0.816
2,Movie,Joseph Williams,Don't Let Me Be Lonely Tonight,0CoSDzoNIKCRs124s9uTVy,3,0.952,0.663,170267,0.131,0.000,C,0.1030,-13.879,Minor,0.0362,99.488,5/4,0.368
3,Movie,Henri Salvador,Dis-moi Monsieur Gordon Cooper,0Gc6TVm52BwZD07Ki6tIvf,0,0.703,0.240,152427,0.326,0.000,C#,0.0985,-12.178,Major,0.0395,171.758,4/4,0.227
4,Movie,Fabien Nataf,Ouverture,0IuslXpMROHdEPvSl1fTQK,4,0.950,0.331,82625,0.225,0.123,F,0.2020,-21.150,Major,0.0456,140.576,4/4,0.390


In [4]:
df.tail()

,genre,artist_name,track_name,track_id,popularity,acousticness,danceability,duration_ms,energy,instrumentalness,key,liveness,loudness,mode,speechiness,tempo,time_signature,valence
232720,Soul,Slave,Son Of Slide,2XGLdVl7lGeq8ksM6Al7jT,39,0.00384,0.687,326240,0.714,0.544000,D,0.0845,-10.626,Major,0.0316,115.542,4/4,0.962
232721,Soul,Jr Thomas & The Volcanos,Burning Fire,1qWZdkBl4UVPj9lK6HuuFM,38,0.03290,0.785,282447,0.683,0.000880,E,0.2370,-6.944,Minor,0.0337,113.830,4/4,0.969
232722,Soul,Muddy Waters,(I'm Your) Hoochie Coochie Man,2ziWXUmQLrXTiYjCg2fZ2t,47,0.90100,0.517,166960,0.419,0.000000,D,0.0945,-8.282,Major,0.1480,84.135,4/4,0.813
232723,Soul,R.LUM.R,With My Words,6EFsue2YbIG4Qkq8Zr9Rir,44,0.26200,0.745,222442,0.704,0.000000,A,0.3330,-7.137,Major,0.1460,100.031,4/4,0.489
232724,Soul,Mint Condition,You Don't Have To Hurt No More,34XO9RwPMKjbvRry54QzWn,35,0.09730,0.758,323027,0.470,0.000049,G#,0.0836,-6.708,Minor,0.0287,113.897,4/4,0.479


### Dataset Dimensions

Checks the number of rows and columns in the dataset.  
This gives an overview of dataset size and helps estimate computational requirements.


In [5]:
df.shape

(232725, 18)

# Basic Analysis

In [6]:
df.dtypes

genre                object
artist_name          object
track_name           object
track_id             object
popularity            int64
acousticness        float64
danceability        float64
duration_ms           int64
energy              float64
instrumentalness    float64
key                  object
liveness            float64
loudness            float64
mode                 object
speechiness         float64
tempo               float64
time_signature       object
valence             float64
dtype: object

In [7]:
df.describe()

,popularity,acousticness,danceability,duration_ms,energy,instrumentalness,liveness,loudness,speechiness,tempo,valence
count,232725.000000,232725.000000,232725.000000,2.327250e+05,232725.000000,232725.000000,232725.000000,232725.000000,232725.000000,232725.000000,232725.000000
mean,41.127502,0.368560,0.554364,2.351223e+05,0.570958,0.148301,0.215009,-9.569885,0.120765,117.666585,0.454917
std,18.189948,0.354768,0.185608,1.189359e+05,0.263456,0.302768,0.198273,5.998204,0.185518,30.898907,0.260065
min,0.000000,0.000000,0.056900,1.538700e+04,0.000020,0.000000,0.009670,-52.457000,0.022200,30.379000,0.000000
25%,29.000000,0.037600,0.435000,1.828570e+05,0.385000,0.000000,0.097400,-11.771000,0.036700,92.959000,0.237000
50%,43.000000,0.232000,0.571000,2.204270e+05,0.605000,0.000044,0.128000,-7.762000,0.050100,115.778000,0.444000
75%,55.000000,0.722000,0.692000,2.657680e+05,0.787000,0.035800,0.264000,-5.501000,0.105000,139.054000,0.660000
max,100.000000,0.996000,0.989000,5.552917e+06,0.999000,0.999000,1.000000,3.744000,0.967000,242.903000,1.000000


In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 232725 entries, 0 to 232724
Data columns (total 18 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   genre             232725 non-null  object 
 1   artist_name       232725 non-null  object 
 2   track_name        232724 non-null  object 
 3   track_id          232725 non-null  object 
 4   popularity        232725 non-null  int64  
 5   acousticness      232725 non-null  float64
 6   danceability      232725 non-null  float64
 7   duration_ms       232725 non-null  int64  
 8   energy            232725 non-null  float64
 9   instrumentalness  232725 non-null  float64
 10  key               232725 non-null  object 
 11  liveness          232725 non-null  float64
 12  loudness          232725 non-null  float64
 13  mode              232725 non-null  object 
 14  speechiness       232725 non-null  float64
 15  tempo             232725 non-null  float64
 16  time_signature    23

Converting ms to seconds for ease

In [9]:
df['duration_ms']=round(df['duration_ms']/1000,2)
df.rename(columns={'duration_ms':'duration_sec'},inplace=True)
df.head()

,genre,artist_name,track_name,track_id,popularity,acousticness,danceability,duration_sec,energy,instrumentalness,key,liveness,loudness,mode,speechiness,tempo,time_signature,valence
0,Movie,Henri Salvador,C'est beau de faire un Show,0BRjO6ga9RKCKjfDqeFgWV,0,0.611,0.389,99.37,0.910,0.000,C#,0.3460,-1.828,Major,0.0525,166.969,4/4,0.814
1,Movie,Martin & les fées,Perdu d'avance (par Gad Elmaleh),0BjC1NfoEOOusryehmNudP,1,0.246,0.590,137.37,0.737,0.000,F#,0.1510,-5.559,Minor,0.0868,174.003,4/4,0.816
2,Movie,Joseph Williams,Don't Let Me Be Lonely Tonight,0CoSDzoNIKCRs124s9uTVy,3,0.952,0.663,170.27,0.131,0.000,C,0.1030,-13.879,Minor,0.0362,99.488,5/4,0.368
3,Movie,Henri Salvador,Dis-moi Monsieur Gordon Cooper,0Gc6TVm52BwZD07Ki6tIvf,0,0.703,0.240,152.43,0.326,0.000,C#,0.0985,-12.178,Major,0.0395,171.758,4/4,0.227
4,Movie,Fabien Nataf,Ouverture,0IuslXpMROHdEPvSl1fTQK,4,0.950,0.331,82.62,0.225,0.123,F,0.2020,-21.150,Major,0.0456,140.576,4/4,0.390


### Missing Value Analysis

Identifies null or missing values across all columns.  
Ensures data quality before applying transformations and similarity computations.
Also checking if duplicates exist.

In [10]:
df.isnull().sum()

genre               0
artist_name         0
track_name          1
track_id            0
popularity          0
acousticness        0
danceability        0
duration_sec        0
energy              0
instrumentalness    0
key                 0
liveness            0
loudness            0
mode                0
speechiness         0
tempo               0
time_signature      0
valence             0
dtype: int64

In [11]:
df.dropna(inplace=True)

In [12]:
df.isnull().sum()

genre               0
artist_name         0
track_name          0
track_id            0
popularity          0
acousticness        0
danceability        0
duration_sec        0
energy              0
instrumentalness    0
key                 0
liveness            0
loudness            0
mode                0
speechiness         0
tempo               0
time_signature      0
valence             0
dtype: int64

In [13]:
df.duplicated().sum()

np.int64(0)


splitting up the categorical and numerical columns

In [14]:
cat_cols=df.select_dtypes(include=object).columns
num_cols=df.select_dtypes(include=np.number).columns

In [15]:
for col in cat_cols:
    print(df[col].value_counts())

genre
Comedy              9681
Soundtrack          9646
Indie               9543
Jazz                9441
Pop                 9386
Electronic          9377
Children’s Music    9353
Folk                9299
Hip-Hop             9295
Rock                9272
Alternative         9263
Classical           9256
Rap                 9232
World               9095
Soul                9089
Blues               9023
R&B                 8992
Anime               8936
Reggaeton           8927
Ska                 8874
Reggae              8771
Dance               8701
Country             8664
Opera               8280
Movie               7806
Children's Music    5403
A Capella            119
Name: count, dtype: int64
artist_name
Giuseppe Verdi            1394
Giacomo Puccini           1137
Kimbo Children's Music     971
Nobuo Uematsu              825
Richard Wagner             804
                          ... 
Major Harris                 1
The Magnolia                 1
Scienze                      1
De

### Aggregation

aggregating track ids together to prevent multiple entries and biasness

Duration is excluded as it reflects song length rather than musical style or mood, and would add noise to audio-feature-based similarity.

In [16]:
df = df.groupby('track_id', as_index=False).agg({
    'track_name': 'first',
    'artist_name': 'first',
    'genre': 'first',
    'popularity': 'mean',
    'acousticness': 'mean',
    'danceability': 'mean',
    'energy': 'mean',
    'instrumentalness': 'mean',
    'liveness': 'mean',
    'loudness': 'mean',
    'speechiness': 'mean',
    'tempo': 'mean',
    'valence': 'mean',
    'key': lambda x: x.mode()[0],
    'mode': lambda x: x.mode()[0],
    'time_signature': lambda x: x.mode()[0]
})

In [17]:
df['track_id'].value_counts()

track_id
7zzbfi8fvHe6hm342GcNYl    1
00021Wy6AyMbLP2tqij86e    1
000CzNKC8PEt1yC3L8dqwV    1
000DfZJww8KiixTKuk9usJ    1
000EWWBkYaREzsBplYjUag    1
                         ..
001YQlnDSduXd5LgBd66gT    1
001VMKfkHZrlyj7JlQbQFL    1
001KkOBeRiQ1J7IEJYHODW    1
001CyR8xqmmpVZFiTZJ5BC    1
0017XiMkqbTfF2AUOzlhj6    1
Name: count, Length: 176773, dtype: int64

### Extracting meta data

These features dont contribute in building the recommendation system.
So separating features like track names, artist names, and other descriptive metadata.  
This metadata is later used to display human-readable recommendations.

In [18]:
meta=df[['track_id','track_name','artist_name']]

### Removing Non-Useful Columns from main dataframe

Removes columns that do not contribute to content-based similarity, such as identifiers or redundant metadata.  
This simplifies the feature space.


In [19]:
df.drop(['track_id','track_name','artist_name'],axis=1,inplace=True)

In [20]:
cat_cols=df.select_dtypes(include=object).columns
num_cols=df.select_dtypes(include=np.number).columns

In [21]:
for col in cat_cols:
    print(df[col].value_counts())

genre
Comedy              9674
Electronic          9149
Alternative         9095
Anime               8935
Classical           8711
Reggae              8687
Reggaeton           8549
Blues               8496
Soundtrack          8480
Opera               8280
Folk                8048
Jazz                8039
Ska                 8017
Dance               7982
World               7972
Movie               7802
Hip-Hop             7413
Country             7383
Children's Music    5402
R&B                 5353
Soul                4430
Indie               3318
Pop                 2417
Rock                2227
Rap                 1456
Children’s Music    1339
A Capella            119
Name: count, dtype: int64
key
C     20970
G     20476
D     18643
A     17499
C#    16856
F     15604
E     13408
B     12893
A#    11924
F#    11365
G#    11175
D#     5960
Name: count, dtype: int64
mode
Major    116618
Minor     60155
Name: count, dtype: int64
time_signature
4/4    149186
3/4     20806
5/4      4430

# Preprocessing

### Encoding Categorical Features

Encodes categorical attributes such as artists or genres into numerical form if required.  
Ensures compatibility with mathematical similarity methods.
Made use of ***One Hot Encoding*** because number of classes or not very high.

In [22]:
from sklearn.preprocessing import OneHotEncoder

ohe=OneHotEncoder(sparse_output=False)
cat_cols_encoded=ohe.fit_transform(df[cat_cols])
encoded_cols=ohe.get_feature_names_out()
cat_cols_encoded=pd.DataFrame(cat_cols_encoded,columns=encoded_cols,index=df.index)
cat_cols_encoded

,genre_A Capella,genre_Alternative,genre_Anime,genre_Blues,genre_Children's Music,genre_Children’s Music,genre_Classical,genre_Comedy,genre_Country,genre_Dance,...,key_F#,key_G,key_G#,mode_Major,mode_Minor,time_signature_0/4,time_signature_1/4,time_signature_3/4,time_signature_4/4,time_signature_5/4
0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,1.0,0.0,1.0,0.0,0.0,0.0,0.0,1.0,0.0
1,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,1.0,0.0
2,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,1.0,0.0
3,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,1.0,0.0
4,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,...,0.0,1.0,0.0,1.0,0.0,0.0,0.0,0.0,1.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
176768,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,1.0,0.0
176769,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,1.0,1.0,0.0,0.0,0.0,1.0,0.0,0.0
176770,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,1.0,0.0,0.0,1.0,0.0,0.0,0.0,1.0,0.0
176771,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,1.0,0.0


In [23]:
df[num_cols]

,popularity,acousticness,danceability,energy,instrumentalness,liveness,loudness,speechiness,tempo,valence
0,13.0,0.2340,0.617,0.862,0.976000,0.1410,-12.855,0.0514,129.578,0.886
1,5.0,0.2490,0.518,0.805,0.000000,0.3330,-6.248,0.0407,79.124,0.841
2,30.0,0.3660,0.631,0.513,0.000004,0.1090,-6.376,0.0293,120.365,0.307
3,39.0,0.8150,0.768,0.137,0.922000,0.1130,-13.284,0.0747,76.430,0.560
4,70.0,0.1310,0.748,0.627,0.000000,0.0852,-6.029,0.0644,120.963,0.524
...,...,...,...,...,...,...,...,...,...,...
176768,38.0,0.4000,0.851,0.790,0.230000,0.1230,-4.641,0.0309,111.993,0.889
176769,40.0,0.9730,0.538,0.144,0.177000,0.3590,-17.711,0.0569,135.640,0.110
176770,28.0,0.1130,0.716,0.806,0.000000,0.1510,-5.913,0.1350,136.069,0.544
176771,67.0,0.2170,0.664,0.537,0.000003,0.1180,-6.000,0.0290,115.049,0.221


### Checking for skewness in numerical features

Checking if numerical columns are skewed, so that appropriate transformation can be applied.

In [24]:
df[num_cols].skew()

popularity         -0.126046
acousticness        0.366474
danceability       -0.332193
energy             -0.330868
instrumentalness    1.578591
liveness            1.956987
loudness           -1.475308
speechiness         3.102421
tempo               0.406745
valence             0.148548
dtype: float64

Seperating skewed columns, i.e, the columns with absolute skew value greater than 1

In [25]:
skewed_cols=[]
for col in num_cols:
    if abs(df[col].skew())>1:
        skewed_cols.append(col)
skewed_cols

['instrumentalness', 'liveness', 'loudness', 'speechiness']

### Transformation

Used automated power transformer class from scikit learn library with yeo johnson method, so that it can handle positive, negative and zero values efficiently and transform them.

In [26]:
from sklearn.preprocessing import PowerTransformer

pt=PowerTransformer(method='yeo-johnson')
df_transformed=pt.fit_transform(df[skewed_cols])
df_transformed=pd.DataFrame(df_transformed,columns=skewed_cols,index=df[num_cols].index)
df_transformed.head()

,instrumentalness,liveness,loudness,speechiness
0,1.869703,-0.181602,-0.681878,-0.372198
1,-0.619305,1.132003,0.532260,-0.655848
2,-0.619233,-0.584625,0.499248,-0.997398
3,1.864849,-0.529733,-0.738633,0.141979
4,-0.619305,-0.941716,0.590146,-0.069418


Seperating non skewed columns 

In [27]:
non_skewed=set(num_cols)-set(skewed_cols)
non_skewed=list(non_skewed)

### Final numerical data preparation

In [28]:
num_data=pd.concat([df_transformed,df[non_skewed]],axis=1)

### Feature Scaling

Applies normalization/standardization to ensure all audio features are on a comparable scale.  
This prevents features with large ranges from dominating similarity calculations.

Used **MinMax Scaler** because recommendation systems work based on distances.

In [29]:
from sklearn.preprocessing import MinMaxScaler

mms=MinMaxScaler()
num_data_scaled=mms.fit_transform(num_data)
num_data_scaled=pd.DataFrame(num_data_scaled,columns=num_data.columns,index=num_data.index)


### Final Feature Dataset Assembly

Merges numerical audio features and encoded categorical data into a single dataset.  
This dataset fully represents each song for recommendation purposes.


In [30]:
df_final=pd.concat([cat_cols_encoded,num_data_scaled],axis=1)
df_final.tail()

,genre_A Capella,genre_Alternative,genre_Anime,genre_Blues,genre_Children's Music,genre_Children’s Music,genre_Classical,genre_Comedy,genre_Country,genre_Dance,...,instrumentalness,liveness,loudness,speechiness,valence,popularity,danceability,acousticness,energy,tempo
176768,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.766061,0.479840,0.143270,0.085348,0.889,0.38,0.851947,0.401606,0.790787,0.384023
176769,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.680154,0.845906,0.066075,0.296338,0.110,0.40,0.516146,0.976908,0.144127,0.495290
176770,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.000000,0.553623,0.130047,0.668038,0.544,0.28,0.707113,0.113454,0.806803,0.497309
176771,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,...,0.000020,0.465265,0.129236,0.067413,0.221,0.67,0.651325,0.217871,0.537528,0.398402
176772,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.075084,0.308263,0.104338,0.101076,0.669,0.33,0.510782,0.030020,0.547538,0.198222


# Model Training

### Sparse Matrix Optimization

Converts the final feature matrix into a sparse format.  
This reduces memory usage and improves performance during similarity computations.


In [31]:
from scipy.sparse import csr_matrix
df_final_csr=csr_matrix(df_final)

### Similarity Metric Setup

Prepares the dataset for cosine similarity calculations.  
Cosine similarity measures how similar two songs are based on feature direction rather than magnitude.
This function outputs the **Cosine Similarity matrix** with which the model picks the suggested songs.

In [32]:
meta[meta['track_name']=='Houdini']

,track_id,track_name,artist_name
119426,5Fli1xRi01bvCjsZvKWro0,Houdini,Foster The People
168159,7cc6o0TiaDVvMAwZt5STd5,Houdini,"nothing,nowhere."


In [33]:
from sklearn.metrics.pairwise import cosine_similarity
def recommendation_matrix(song,artist,songs_data):
    song_row=meta[(meta['track_name'].str.lower()==song.lower()) & (meta['artist_name'].str.lower()==artist.lower())]
    song_idx=song_row.index[0]
    if not song_row.empty:
        similarity_matrix=cosine_similarity(songs_data[song_idx],songs_data).flatten()
        return similarity_matrix,song_idx
    return "Song doesnt Exist",-1
      

### Recommendation Logic

Defines the core logic for identifying similar songs based on cosine similarity scores.  
The model ranks songs and excludes the input song itself from results.


### Content-Based Similarity Model

Constructs the similarity model using cosine similarity across all tracks.  
This model forms the core of the recommendation system and consists of two functions for two seperate purposes.

The ***content-based*** function returns songs similar in audio features, while the ***artist-based function*** ensures recommendations are restricted to the same artist. For example, querying ‘Night Changes’ produces top recommendations from One Direction in the artist-based approach, while the content-based function may include similar songs from other artists. This dual approach allows flexibility for users seeking either same-artist recommendations or similar-sounding songs. Therecommendation results are formatted into a clean, user-readable list.  
Improves presentation and usability.


#### Content Based Function

In [34]:
def top5_recommendations(song,artist,songs_data):
    final_show=[]
    similar,idx=recommendation_matrix(song,artist,songs_data)
    if isinstance(similar,str):
        return "Song Not Found"
    top5=similar.argsort()[::-1]
    top5=top5[top5!=idx][:5]
    top5_songs=meta.iloc[top5][['track_name','artist_name']].values
    for song in top5_songs:
        final_show.append(song[0]+" "+"by"+" "+song[1])
    return final_show,top5


#### Artist Based Function

In [35]:
def recommend_top5_song_by_artist(song, artist,songs_data, meta):
    final_show=[]
    similar,idx=recommendation_matrix(song,artist,songs_data)
    if isinstance(similar,str):
        return "Song Not Found"
    same_artist_mask = (meta['artist_name'].str.lower().values==artist.lower())
    similar[~same_artist_mask] = -1
    top5=similar.argsort()[::-1][1:6]
    top5=top5[top5!=idx][:5]
    top5_songs=meta.iloc[top5][['track_name','artist_name']].values
    for song in top5_songs:
        final_show.append(song[0]+" "+"by"+" "+song[1])
    return final_show,top5

# Test Case Validation

### Test Case: Song-Based Recommendation

Tests the recommendation system using a popular song as input.  
Verifies similarity accuracy and output format.


In [36]:
print(top5_recommendations("perfect","ed sheeran",df_final_csr)[0])

['do re mi (feat. Gucci Mane) by blackbear', 'Perfect Symphony (Ed Sheeran & Andrea Bocelli) by Ed Sheeran', 'Satellite by Dave Matthews Band', 'Loco Enamorado by Remmy Valenzuela', 'Perfect Duet (Ed Sheeran & Beyoncé) by Ed Sheeran']


### Test Case: Artist-Aware Recommendation

Tests recommendations when both song name and artist are provided.  
Confirms correct disambiguation and relevance.


In [43]:
print(recommend_top5_song_by_artist("perfect","ed sheeran",df_final_csr,meta)[0])

['Perfect Symphony (Ed Sheeran & Andrea Bocelli) by Ed Sheeran', 'Perfect Duet (Ed Sheeran & Beyoncé) by Ed Sheeran', 'Dive by Ed Sheeran', "Hearts Don't Break Around Here by Ed Sheeran", 'Tenerife Sea by Ed Sheeran']


### Test Case: Song based and Artist-Aware Recommendation

Tests recommendations when both song name and artist are provided.  
Confirms correct disambiguation and relevance. Now picking a song with very low popularity to check if its output is reliable.

picked a random song in the below list with index 184 with popularity 0

In [55]:
df[df['popularity']==0]

,genre,popularity,acousticness,danceability,energy,instrumentalness,liveness,loudness,speechiness,tempo,valence,key,mode,time_signature
32,Classical,0.0,0.983000,0.420,0.191,0.889000,0.3600,-18.023,0.0473,137.812,0.0889,F,Major,4/4
64,Classical,0.0,0.876000,0.366,0.196,0.000000,0.4310,-21.798,0.1410,101.237,0.2670,G,Major,4/4
147,Children's Music,0.0,0.985000,0.262,0.264,0.881000,0.0978,-15.362,0.0358,109.331,0.4240,E,Minor,4/4
183,Movie,0.0,0.975000,0.265,0.112,0.577000,0.1190,-20.070,0.0388,79.465,0.0579,G,Major,4/4
184,Children's Music,0.0,0.017100,0.635,0.880,0.255000,0.0529,-8.542,0.0419,120.014,0.2730,A#,Minor,4/4
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
176649,Movie,0.0,0.042700,0.813,0.813,0.000312,0.0499,-5.622,0.0440,135.979,0.9670,G#,Major,4/4
176692,Children's Music,0.0,0.781000,0.582,0.189,0.000000,0.1380,-18.159,0.7590,172.643,0.7910,G,Major,4/4
176742,Dance,0.0,0.000883,0.523,0.610,0.000023,0.2030,-5.864,0.0609,159.001,0.4550,C,Major,3/4
176755,Movie,0.0,0.176000,0.859,0.748,0.000004,0.0909,-8.296,0.0573,120.103,0.9710,A#,Major,4/4


In [58]:
demo=meta.iloc[184,:]

In [59]:
print(recommend_top5_song_by_artist(demo['track_name'],demo['artist_name'],df_final_csr,meta)[0])

['Interruption Please - Instrumental by Children Songs Company', 'Roll - Instrumental by Children Songs Company', 'Sober One - Instrumental by Children Songs Company', 'Gladiolus - Instrumental by Children Songs Company', 'Tomorrow by Children Songs Company']


In [60]:
print(top5_recommendations(demo['track_name'],demo['artist_name'],df_final_csr)[0])

['Interruption Please - Instrumental by Children Songs Company', 'Roll - Instrumental by Children Songs Company', 'Sober One - Instrumental by Children Songs Company', 'Gladiolus - Instrumental by Children Songs Company', 'Dance Jump by Children Songs Company']


For zero-popularity songs, content-based and artist-based recommendations converge, indicating the artist has no other tracks with similar audio features in the dataset. This is expected behavior for niche/obscure tracks.

# Evaluation

### Baseline Comparison

Baseline comparison is used to verify that the recommendation system performs better than random chance. The average cosine similarity of model-generated recommendations is compared against randomly selected songs. Self-similarity is explicitly removed to ensure fair evaluation. Higher similarity scores than the baseline confirm meaningful recommendations.


In [38]:
def avg_similarity(idx, rec_indices, sim_matrix):
    if isinstance(rec_indices, (int, np.integer)):
        rec_indices = [rec_indices]      
    rec_indices = [i for i in rec_indices if i != idx]
    return np.mean(sim_matrix[rec_indices])

evaluating content based model

In [39]:
import random 
random_recs = random.sample(range(df_final_csr.shape[0]), 10) 
sim_matrix1,idx1=recommendation_matrix("Perfect","ed sheeran",df_final_csr) 
avg_model=avg_similarity(idx1,top5_recommendations("Perfect","ed sheeran",df_final_csr)[1],sim_matrix1)  
avg_random=avg_similarity(idx1,random_recs,sim_matrix1) 
print(avg_model,avg_random)

0.9723106972392321 0.34766113666917037


evaluating artist based model

In [40]:
sim_matrix2,idx2=recommendation_matrix("IDGAF","Dua Lipa",df_final_csr) 
avg_model2=avg_similarity(idx2,top5_recommendations("IDGAF","Dua lipa",df_final_csr)[1],sim_matrix2)  
avg_random=avg_similarity(idx2,random_recs,sim_matrix2) 
print(avg_model2,avg_random)

0.9933914829579746 0.538403724768098


Content-based recommendations show a very high similarity (0.972) compared to the random baseline (0.391), indicating strong feature alignment. Artist-based recommendations improve over random as well (0.455 vs 0.391) but are noticeably weaker than full content-based similarity, which is still valid because there is no need an artist performs songs with similar music features.

### Leave One Out Strategy Test

This evaluation uses a leave-one-out strategy where, for each sampled song, the average cosine similarity of its top-k recommendations is compared against a random baseline. By repeating this across multiple songs and averaging the scores, the method verifies that the model consistently produces more similar recommendations than chance.

In [41]:
def evaluate_model(song_indices, k=5):
    model_scores = []
    random_scores = []
    for idx in song_indices:
        sim = cosine_similarity(df_final_csr[idx], df_final_csr).flatten()
        topk = sim.argsort()[::-1]
        topk = [i for i in topk if i != idx][:k]
        randk = random.sample(
            [i for i in range(len(sim)) if i != idx],
            k
        )
        model_scores.append(np.mean(sim[topk]))
        random_scores.append(np.mean(sim[randk]))
    return np.mean(model_scores), np.mean(random_scores)

In [42]:
sample_indices = random.sample(range(df_final_csr.shape[0]), 50)
evaluate_model(sample_indices)

(np.float64(0.9918678090676866), np.float64(0.5073636733922247))

This score validates that the recommendation system computes correct cosine similarities and the recommendations provided are valid when compared to a random suggestor.

### Conclusion
  
All components of the content-based Spotify recommendation system have been successfully implemented and tested.
Popularity is retained as a supporting feature but given equal weight during scaling. Future work could include popularity-aware re-ranking on top of cosine similarity scores for more balanced recommendations."
